File: InterleaveBits.py
def InterleaveChunk(inchunk, numrows, numcols):
	'''Interleave the input data. You shouldn't be calling this function manually.'''
	
	if len(inchunk) != (numrows * numcols):
		raise RuntimeError("The input chunk doesn't fit in the interleave matrix.")
	i = 0
	rownum = 0
	colnum = 0
	temp = np.zeros(x.shape)
	out = np.zeros(x.shape)

	while i < numrows * numcols:
		count = 0
		while count < numrows:
			temp[colnum + numcols * rownum] = inchunk[i]
			rownum = (rownum + 7) % numrows
			count += 1
			i += 1
		rownum = 0
		colnum += 1

	i = 0
	rownum = 0
	colnum = 0
	del count

	while i < numrows * numcols:
		lastcolnum = colnum
		while rownum < numrows:
			out[i] = temp[colnum + numcols * rownum]
			rownum += 1
			colnum = (colnum - 7) % numcols
			i += 1
		rownum = 0
		colnum = lastcolnum + 1
		lastcolnum = colnum
	return out

def InterleaveData(indata, Bd=75, interleave_len="S"):
	'''Chunkwise interleave the input data.'''
	if interleave_len == "Z":
		return inchunk
	elif Bd == 75 and interleave_len == "S":
		numrows = 10
		numcols = 9
	elif Bd == 75 and interleave_len == "L":
		numrows = 20
		numcols = 36
	elif (Bd == 150 or Bd == 300 or Bd == 600) and interleave_len == "S":
		numrows = 40
		numcols = 18
	elif (Bd == 150 or Bd == 300 or Bd == 600) and interleave_len == "L":
		numrows = 40
		numcols = 144
	elif Bd == 1200 and interleave_len == "S":
		numrows = 40
		numcols = 36
	elif Bd == 1200 and interleave_len == "L":
		numrows = 40
		numcols = 288
	elif Bd == 2400 and interleave_len == "S":
		numrows = 40
		numcols = 72
	elif Bd == 2400 and interleave_len == "L":
		numrows = 40
		numcols = 576
	else:
		raise RuntimeError("Invalid input.")
	
	z = numrows * numcols
	
	out = np.zeros(indata.shape)
	
	if (len(indata) % z) != 0:
		raise RuntimeError("The input data does not evenly fit in the interleaver. Something is VERY wrong.")
	
	for i in range(0,len(indata),z):
		temp = indata[i:i+z]
		temp = InterleaveChunk(temp,numrows,numcols)
		out[i:i+z] = temp
	return out

File: interleaveBits.cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include "Types.h"
#include "interleaveBits.h"

void interleave_chunk(std::vector<byte>* chunk, size_t numrows, size_t numcols, byte offset)
{
	if (chunk->size() != ((numcols * numrows) / 8))
	{
		std::cerr << "The input chunk doesn't fit in the interleave matrix. How did you do this?\n\n";
		exit(1);
	}

	size_t i = 0, rownum = 0, colnum = 0;
	std::vector<byte> temp(chunk->size());

	while (i < (numrows * numcols))
	{
		size_t count = 0;

		while (count < numrows)
		{
			setBitVal(&temp, (colnum + numcols * rownum), getBitVal(*chunk, i + offset));
			rownum = (rownum + 7) % numrows;
			count++; i++;
		}

		rownum = 0;
		colnum++;
	}

	i = 0;
	rownum = 0;
	colnum = 0;
	size_t lastcolnum;

	while (i < (numcols * numrows))
	{
		lastcolnum = colnum;

		while (rownum < numrows)
		{
			setBitVal(chunk, i + offset, getBitVal(temp, (colnum + numcols * rownum)));
			rownum++;
			colnum = (colnum - 7) % numcols;
			i++;
		}

		rownum = 0;
		colnum = lastcolnum + 1;
	}
	return;
}

void interleave_data(std::vector<byte>* datastream, size_t baud, interleave_len chunklen, size_t bitlen)
{
	size_t numrows, numcols;
	byte tempmask[2];
	byte offset;

	switch (baud)
	{
	case 150:
	case 300:
	case 600:
	{
		switch (chunklen)
		{
		case _zero:
			return;

		case _short:
			numcols = 18;
			numrows = 40;
			break;

		case _long:
			numcols = 144;
			numrows = 40;
			break;
		}
		break;
	}

	case 75:
	{
		switch (chunklen)
		{
		case _zero:
			return;

		case _short:
			numcols = 9;
			numrows = 10;
			break;

		case _long:
			numcols = 36;
			numrows = 20;
			break;
		}
		break;
	}

	case 1200:
	{
		switch (chunklen)
		{
		case _zero:
			return;

		case _short:
			numcols = 36;
			numrows = 40;
			break;

		case _long:
			numcols = 288;
			numrows = 40;
			break;
		}
		break;
	}

	case 2400:
	{
		switch (chunklen)
		{
		case _zero:
			return;

		case _short:
			numcols = 72;
			numrows = 40;
			break;

		case _long:
			numcols = 576;
			numrows = 40;
			break;
		}
		break;
	}

	default:
		return;
	}

	size_t z = numcols * numrows;

	if ((bitlen % z) != 0)
	{
		std::cerr << "The input data does not fit evenly into the interleaver. Check upstream! (interleave_data)\n\n";
		exit(1);
	}

	std::vector<byte> temp((z / 8) + ((z % 8 != 0) ? 1 : 0));

	for (size_t i = 0; i < bitlen; i += z)
	{
		temp = std::vector<byte>(datastream->begin() + (i / 8), datastream->begin() + ((i + z) / 8));
		offset = (i % 8);

		temp[0] &= ((i % 8 == 0) ? 0xff : ~(0x80 >> ((i % 8) - 1)));
		temp[(z / 8) - 1] &= ((i % 8 == 0) ? 0xff : (0x80 >> ((i & 8) - 1)));

		interleave_chunk(&temp, numrows, numcols, offset);

		tempmask[0] = datastream->at((i / 8)) & ((i % 8 == 0) ? 0xff : (0x80 >> ((i & 8) - 1)));
		tempmask[1] = datastream->at((i + z) / 8) & (((i + z) % 8 == 0) ? 0xff : ~(0x80 >> (((i + z) % 8) - 1)));

		temp[0] |= tempmask[0];
		temp[(z / 8) - 1] |= tempmask[1];

		std::copy(std::begin(temp), std::end(temp), std::begin(*datastream) + i);
	}

	return;
}
